src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+        private static IMethodSymbol GetContainingSymbol(IInvocationOperation invocation, IMethodSymbol containingSymbol)
+        {
+            IOperation currentOperation = invocation.Parent;
+
+            while (currentOperation != null)
+            {
+                if (currentOperation.Kind == OperationKind.AnonymousFunction)
+                {
+                    return ((IAnonymousFunctionOperation)currentOperation).Symbol;
+                }
+                else if (currentOperation.Kind == OperationKind.LocalFunction)
+                {
+                    return ((ILocalFunctionOperation)currentOperation).Symbol;
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return containingSymbol;
+        }
+
+        // Check if the method only takes one ct and is the last parameter in the method signature.
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType)) // Covers the case when using an alias for ct
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+                return true;
+            }
+
+            cancellationTokenParameterName = null;
+            return false;
+        }
+
+        // Checks if the invocation has an optional ct argument at the end or a params ct array at the end.
+        private static bool InvocationMethodTakesAToken(IMethodSymbol method, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            return
+                !method.Parameters.IsEmpty &&
+                method.Parameters[method.Parameters.Length - 1] is IParameterSymbol lastParameter &&
+                (InvocationIgnoresOptionalCancellationToken(lastParameter, arguments, cancellationTokenType) ||
+                InvocationIsUsingParamsCancellationToken(lastParameter, arguments, cancellationTokenType));
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
```

Comment:
Nit: Rename?
```suggestion
        private static bool InvocationUsesOptionalCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
```

Reply:
The current name `InvocationUsesOptionalCancellationToken` is the correct name because it checks if the method is being called with the token default parameter being ignored.